File: DinnerApp/src/lib/leftovers.ts
import { DAY_KEYS, type DayKey, type Meal, type Week } from '../types';

export type LeftoverTile = {
  id: string; // dnd id: `leftover:<mealId>`
  sourceMealId: string;
  mealName: string;
  earliestFromIndex: number; // 0 = Monday
};

/** For each home-cooked meal assigned to a day, expose a leftover tile usable
 *  on any later day. No cascade: leftover-of-leftover is not generated. */
export function deriveLeftoverTiles(week: Week, meals: Meal[]): LeftoverTile[] {
  const byId = new Map(meals.map((m) => [m.id, m]));
  const earliest = new Map<string, number>();
  DAY_KEYS.forEach((k, i) => {
    const a = week[k].assignment;
    if (a && a.kind === 'meal') {
      const prev = earliest.get(a.mealId);
      if (prev === undefined || i < prev) earliest.set(a.mealId, i);
    }
  });
  const tiles: LeftoverTile[] = [];
  earliest.forEach((dayIndex, mealId) => {
    const m = byId.get(mealId);
    if (!m) return;
    tiles.push({
      id: `leftover:${mealId}`,
      sourceMealId: mealId,
      mealName: m.name,
      earliestFromIndex: dayIndex + 1, // strictly later
    });
  });
  return tiles.sort((a, b) => a.mealName.localeCompare(b.mealName));
}

export const dayIndex = (d: DayKey) => DAY_KEYS.indexOf(d);

File: DinnerApp/ios/DinnerPlanner/DinnerPlanner/Services/Leftovers.swift
import Foundation

/// A leftover option derived (not persisted) from the current week.
/// Ported from the web `LeftoverTile` / `deriveLeftoverTiles` (leftovers.ts).
struct LeftoverTile: Identifiable, Hashable {
    var id: String { "leftover:\(sourceMealId)" }
    let sourceMealId: String
    let mealName: String
    /// Earliest day index (0=Mon) this leftover may be used on — strictly after
    /// the day its source meal is cooked.
    let earliestFromIndex: Int
}

enum Leftovers {
    /// For each home-cooked meal assigned to a day, expose a leftover usable on
    /// any *later* day. No cascade (leftover-of-leftover is not generated).
    static func derive(days: [DayPlan], meals: [Meal]) -> [LeftoverTile] {
        let byId = Dictionary(uniqueKeysWithValues: meals.map { ($0.id, $0) })

        // Earliest day index each meal is cooked on.
        var earliest: [String: Int] = [:]
        for day in days {
            guard day.assignmentKind == .meal, let mealId = day.assignmentRefId else { continue }
            let i = day.dayIndex
            if let prev = earliest[mealId] {
                if i < prev { earliest[mealId] = i }
            } else {
                earliest[mealId] = i
            }
        }

        var tiles: [LeftoverTile] = []
        for (mealId, dayIdx) in earliest {
            guard let meal = byId[mealId] else { continue }
            tiles.append(LeftoverTile(sourceMealId: mealId,
                                      mealName: meal.name,
                                      earliestFromIndex: dayIdx + 1)) // strictly later
        }
        return tiles.sorted { $0.mealName.localizedCaseInsensitiveCompare($1.mealName) == .orderedAscending }
    }
}
